activemodel/lib/active_model/attribute_registration.rb

```diff
@@ -40,6 +40,10 @@ def attribute_types # :nodoc:
         end
       end
 
+      def type_for_attribute(attribute)
+        attribute_types[attribute.to_s]
+      end
```

Comment:
Defining `type_for_attribute` like this here will shadow the [definition in `ActiveRecord::ModelSchema`](https://github.com/rails/rails/blob/d0f40f71b0b61f828291f366bad5d29ab44e70d2/activerecord/lib/active_record/model_schema.rb#L453-L462) because `ActiveRecord::Attributes` includes `ActiveModel::AttributeRegistration`, and it is included after `ActiveRecord::ModelSchema`.

I think it's worth splitting this out into its own PR, including tests, so I submitted #49910 and added you as a co-author.

Reply:
Nice! Thanks @jonathanhefner